File: src/api/go/fog05-go/fdu.go
/*
* Copyright (c) 2014,2019 Contributors to the Eclipse Foundation
* See the NOTICE file(s) distributed with this work for additional
* information regarding copyright ownership.
* This program and the accompanying materials are made available under the
* terms of the Eclipse Public License 2.0 which is available at
* http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
* which is available at https://www.apache.org/licenses/LICENSE-2.0.
* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
* Contributors: Gabriele Baldoni, ADLINK Technology Inc.
* golang APIs
 */

package fog05

const (
	LIVE string = "LIVE"
	COLD string = "COLD"

	BARE   string = "BARE"
	KVM    string = "KVM"
	KVMUK  string = "KVM_UK"
	XEN    string = "XEN"
	XENUK  string = "XEN_UK"
	LXD    string = "LXD"
	DOCKER string = "DOCKER"
	MCU    string = "MCU"

	SCRIPT    string = "SCRIPT"
	CLOUDINIT string = "CLOUD_INIT"

	INTERNAL  string = "INTERNAL"
	EXTERNAL  string = "EXTERNAL"
	WLAN      string = "WLAN"
	BLUETOOTH string = "BLUETOOTH"

	PARAVIRT       string = "PARAVIRT"
	FOSMGMT        string = "FOS_MGMT"
	PCIPASSTHROUGH string = "PCI_PASSTHROUGH"
	SRIOV          string = "SR_IOV"
	E1000          string = "E1000"
	RTL8139        string = "RTL8139"
	PHYSICAL       string = "PHYSICAL"
	BRIDGED        string = "BRIDGED"

	GPIO string = "GPIO"
	I2C  string = "I2C"
	BUS  string = "BUS"
	COM  string = "COM"
	CAN  string = "CAN"

	BLOCK  string = "BLOCK"
	FILE   string = "FILE"
	OBJECT string = "OBJECT"

	DEFINE    string = "DEFINE"
	CONFIGURE string = "CONFIGURE"
	CLEAN     string = "CLEAN"
	RUN       string = "RUN"
	STARTING  string = "STARTING"
	STOP      string = "STOP"
	RESUME    string = "RESUME"
	PAUSE     string = "PAUSE"
	SCALE     string = "SCALE"
	TAKEOFF   string = "TAKE_OFF"
	LAND      string = "LAND"
	MIGRATE   string = "MIGRATE"
	UNDEFINE  string = "UNDEFINE"
	ERROR     string = "ERROR"
)

// FDUImage ...
type FDUImage struct {
	UUID     *string `json:"uuid,omitempty"`
	Name     *string `json:"name,omitempty"`
	URI      string  `json:"uri"`
	Checksum string  `json:"checksum"` //SHA1SUM
	Format   string  `json:"format"`
}

// FDUCommand ...
type FDUCommand struct {
	Binary string   `json:"binary"`
	Args   []string `json:"args"`
}

// FDUGeographicalRequirements ...
type FDUGeographicalRequirements struct {
	Position  *FDUPosition  `json:"position,omitempty"`
	Proximity *FDUProximity `json:"proximity,omitempty"`
}

// FDUPosition ...
type FDUPosition struct {
	Latitude  string  `json:"lat"`
	Longitude string  `json:"lon"`
	Radius    float64 `json:"radius"`
}

// FDUProximity ...
type FDUProximity struct {
	Neighbor string  `json:"neighbour"`
	Radius   float64 `json:"radius"`
}

// FDUEnergyRequirements ...
type FDUEnergyRequirements struct {
	Key string `json:"key"`
}

// FDUComputationalRequirements ...
type FDUComputationalRequirements struct {
	Name            *string  `json:"name,omitempty"`
	UUID            *string  `json:"uuid,omitempty"`
	CPUArch         string   `json:"cpu_arch"`
	CPUMinFrequency int      `json:"cpu_min_freq"`
	CPUMinCount     int      `json:"cpu_min_count"`
	GPUMinCount     *int     `json:"gpu_min_count,omitempty"`
	FPGAMinCount    *int     `json:"fpga_min_count,omitempty"`
	RAMSizeMB       float64  `json:"ram_size_mb"`
	StorageSizeGB   float64  `json:"storage_size_gb"`
	DutyCycle       *float64 `json:"duty_cycle,omitempty"`
}

// FDUConfiguration ...
type FDUConfiguration struct {
	ConfType string   `json:"conf_type"`
	Script   string   `json:"script"`
	SSHKeys  []string `json:"ssh_keys,omitempty"`
}

// FDUVirtualInterface ...
type FDUVirtualInterface struct {
	InterfaceType string `json:"intf_type"`
	VPCI          string `json:"vpci"`
	Bandwidth     int    `json:"bandwidth"`
}

// FDUIOPort ...
type FDUIOPort struct {
	Address    string `json:"address"`
	IOKind     string `json:"io_kind"`
	MinIOPorts int    `json:"min_io_ports"`
}

// FDUInterfaceDescriptor ...
type FDUInterfaceDescriptor struct {
	Name             string              `json:"name"`
	IsMGMT           bool                `json:"is_mgmt"`
	InterfaceType    string              `json:"if_type"`
	MACAddress       *string             `json:"mac_address,omitempty"`
	VirtualInterface FDUVirtualInterface `json:"virtual_interface"`
	CPID             *string             `json:"cp_id,omitempty"`
	ExtCPID          *string             `json:"ext_cp_id,omitempty"`
}

// FDUStorageDescriptor ...
type FDUStorageDescriptor struct {
	ID                 string  `json:"id"`
	StorageType        string  `json:"storage_type"`
	Size               int     `json:"size"`
	FileSystemProtocol *string `json:"file_system_protocol,omitempty"`
	CPID               *string `json:"cp_id,omitempty"`
}

// FDU ...
type FDU struct {
	ID                       string                       `json:"id"`
	Name                     string                       `json:"name"`
	UUID                     *string                      `json:"uuid,omitempty"`
	Description              *string                      `json:"description,omitempty"`
	ComputationRequirements  FDUComputationalRequirements `json:"computation_requirements"`
	Image                    *FDUImage                    `json:"image,omitempty"`
	Command                  *FDUCommand                  `json:"command,omitempty"`
	Storage                  []FDUStorageDescriptor       `json:"storage"`
	GeographicalRequirements *FDUGeographicalRequirements `json:"geographical_requirements,omitempty"`
	EnergyRequirements       *FDUEnergyRequirements       `json:"energy_requirements,omitempty"`
	Hypervisor               string                       `json:"hypervisor"`
	MigrationKind            string                       `json:"migration_kind"`
	Configuration            *FDUConfiguration            `json:"configuration,omitempty"`
	Interfaces               []FDUInterfaceDescriptor     `json:"interfaces"`
	IOPorts                  []FDUIOPort                  `json:"io_ports"`
	ConnectionPoints         []ConnectionPointDescriptor  `json:"connection_points"`
	DependsOn                []string                     `json:"depends_on"`
}

// FDUStorageRecord ...
type FDUStorageRecord struct {
	UUID               string  `json:"uuid"`
	StorageID          string  `json:"storage_id"`
	StorageType        string  `json:"storage_type"`
	Size               int     `json:"size"`
	FileSystemProtocol *string `json:"file_system_protocol,omitempty"`
	CPID               *string `json:"cp_id,omitempty"`
}

// FDUInterfaceRecord ...
type FDUInterfaceRecord struct {
	Name                 string               `json:"name"`
	IsMGMT               bool                 `json:"is_mgmt"`
	InterfaceType        string               `json:"if_type"`
	MACAddress           *string              `json:"mac_address,omitempty"`
	VirtualInterface     *FDUVirtualInterface `json:"virtual_interface,omitempty"`
	CPID                 *string              `json:"cp_id,omitempty"`
	ExtCPID              *string              `json:"ext_cp_id,omitempty"`
	VirtualInterfaceName string               `json:"vintf_name"`
	Status               string               `json:"status"`
	PhysicalFace         *string              `json:"phy_face,omitempty"`
	VEthFaceName         *string              `json:"veth_face_name,omitempty"`
	Properties           *jsont               `json:"properties,omitempty"`
}

// FDUMigrationProperties ...
type FDUMigrationProperties struct {
	Destination string `json:"destination"`
	Source      string `json:"source"`
}

// FDURecord ...
type FDURecord struct {
	UUID                     string                       `json:"uuid"`
	FDUID                    string                       `json:"fdu_id"`
	Status                   string                       `json:"status"`
	Image                    *FDUImage                    `json:"image,omitempty"`
	Command                  *FDUCommand                  `json:"command,omitempty"`
	Storage                  []FDUStorageRecord           `json:"storage"`
	ComputationRequirements  FDUComputationalRequirements `json:"computation_requirements"`
	GeographicalRequirements *FDUGeographicalRequirements `json:"geographical_requirements,omitempty"`
	EnergyRequirements       *FDUEnergyRequirements       `json:"energy_requirements,omitempty"`
	Hypervisor               string                       `json:"hypervisor"`
	MigrationKind            string                       `json:"migration_kind"`
	Configuration            *FDUConfiguration            `json:"configuration,omitempty"`
	Interfaces               *[]FDUInterfaceRecord        `json:"interfaces,omitempty"`
	IOPorts                  []FDUIOPort                  `json:"io_ports"`
	ConnectionPoints         []ConnectionPointRecord      `json:"connection_points"`
	DependsOn                []string                     `json:"depends_on"`
	ErrorCode                *int                         `json:"error_code,omitempty"`
	ErrorMsg                 *string                      `json:"error_msg,omitempty"`
	MigrationProperties      *FDUMigrationProperties      `json:"migration_properties,omitempty"`
	HypervisorInfo           *jsont                       `json:"hypervisor_info,omitempty"`
}

File: fog05-sdk/src/im/fdu.rs
/*********************************************************************************
* Copyright (c) 2018,2020 ADLINK Technology Inc.
*
* This program and the accompanying materials are made available under the
* terms of the Eclipse Public License 2.0 which is available at
* http://www.eclipse.org/legal/epl-2.0, or the Apache Software License 2.0
* which is available at https://www.apache.org/licenses/LICENSE-2.0.
*
* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
* Contributors:
*   ADLINK fog05 team, <fog05@adlink-labs.tech>
*********************************************************************************/

#![allow(non_camel_case_types)]

extern crate derive_more;
extern crate serde;

use derive_more::Display;
use semver::Version;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

fn default_zero() -> u8 {
    0
}

fn default_zero_u64() -> u64 {
    0
}

fn default_one() -> u8 {
    1
}

// FDU Generic

// Descriptors
#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "UPPERCASE")]
pub enum MigrationKind {
    LIVE,
    COLD,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "UPPERCASE")]
pub enum ConfigurationKind {
    SCRIPT,
    ENV,
    CLOUD_INIT,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "UPPERCASE")]
pub enum InterfaceKind {
    VIRTUAL,
    WLAN,
    BLUETOOTH,
}

#[derive(Serialize, Deserialize, Debug, Clone, Display)]
#[serde(rename_all = "UPPERCASE")]
pub enum VirtualInterfaceKind {
    PARAVIRT,
    PCI_PASSTHROUGH,
    SR_IOV,
    VIRTIO, //default
    E1000,
    RTL8139,
    PCNET,
    BRIDGED,
    PHYSICAL,
}

#[derive(Serialize, Deserialize, Debug, Clone, Display)]
#[serde(rename_all = "UPPERCASE")]
pub enum StorageKind {
    BLOCK, //virtual disk
    //FILE, //NFS kind of, can be Zenoh+file backend
    OBJECT, //Zenoh as object storage
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "UPPERCASE")]
pub enum ScalingMetric {
    CPU,
    DISK,
    MEMORY,
    CUSTOM(String),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ScalingPolicy {
    pub metric: ScalingMetric,
    pub scale_up_threshold: f32,
    pub scale_down_threshold: f32,
    pub threshold_sensibility: u8,
    pub probe_interval: f32,
    pub min_replicas: u8,
    pub max_replicas: u8,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Position {
    pub lat: String,
    pub lon: String,
    pub radius: f64,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
//or Affinity/Antiaffinity
pub struct Proximity {
    pub neighbour: String,
    pub radius: f64,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Configuration {
    pub conf_kind: ConfigurationKind,
    pub script: Option<String>,   //both for script and cloud_init
    pub env: Option<Vec<String>>, //VAR=VALUE,
    pub ssh_keys: Option<Vec<String>>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Image {
    pub uuid: Option<Uuid>,
    pub name: Option<String>,
    pub uri: String,
    pub checksum: String, //SHA256 of image file
    pub format: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ComputationalRequirements {
    pub cpu_arch: String,
    #[serde(default = "default_zero_u64")]
    pub cpu_min_freq: u64, //default 0 in MHz
    #[serde(default = "default_one")]
    pub cpu_min_count: u8, //default 1
    #[serde(default = "default_zero")]
    pub gpu_min_count: u8, //default 0
    #[serde(default = "default_zero")]
    pub fpga_min_count: u8, //default 0
    pub operating_system: Option<String>,
    pub ram_size_mb: u32,
    pub storage_size_mb: u32,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct GeographicalRequirement {
    pub position: Option<Position>,
    pub proximity: Option<Vec<Proximity>>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct VirtualInterface {
    pub vif_kind: VirtualInterfaceKind,
    pub parent: Option<String>, //PCI address, bridge name, interface name
    pub bandwidht: Option<u8>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ConnectionPointDescriptor {
    pub uuid: Option<Uuid>,
    pub name: String,
    pub id: String,
    pub vld_ref: Option<String>, //reference to a virtual link descriptor
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Interface {
    pub name: String,
    pub kind: InterfaceKind,
    pub mac_address: Option<crate::types::MACAddress>,
    pub virtual_interface: VirtualInterface,
    pub cp_id: Option<String>, //internal to this descriptor
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct StorageDescriptor {
    pub id: String,
    pub storage_kind: StorageKind,
    pub size: u32, //depends on the kind, MB for BLOCK, items for OBJECT
}

//example pf hypervisor for BARE (Native)
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Command {
    pub binary: String, //can be relative, in that case it is expected to be part of the image of global path
    pub args: Vec<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct FDUDescriptor {
    pub uuid: Option<Uuid>,
    pub id: String,
    pub name: String,
    pub version: Version,     //semantic version of the descriptor
    pub fdu_version: Version, //semantic version of the fdu
    pub description: Option<String>,
    pub hypervisor: String, //eg. KVM, LXD, DOCKER, ROS2, BARE, K8s, AWS...
    pub image: Option<Image>,
    pub hypervisor_specific: Option<String>,
    pub computation_requirements: ComputationalRequirements,
    pub geographical_requirements: Option<GeographicalRequirement>,
    pub interfaces: Vec<Interface>,
    pub storage: Vec<StorageDescriptor>,
    pub connection_points: Vec<ConnectionPointDescriptor>,
    pub configuration: Option<Configuration>,
    pub migration_kind: MigrationKind,
    pub replicas: Option<u8>,
    //pub scaling_policies : Option<Vec<ScalingPolicy>>,
    pub depends_on: Vec<String>,
}

// FDU Record

#[derive(Serialize, Deserialize, Debug, Clone, Display)]
#[serde(rename_all = "UPPERCASE")]
pub enum FDUState {
    DEFINED,
    CONFIGURED,
    RUNNING,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct FDURecord {
    pub uuid: Uuid,
    pub fdu_uuid: Uuid,
    pub node: Uuid,
    pub interfaces: Vec<FDURecordInterface>,
    pub connection_points: Vec<FDURecordConnectionPoint>,
    pub status: FDUState,
    pub error: Option<crate::fresult::FError>,
    pub hypervisor_specific: Option<Vec<u8>>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct FDURecordInterface {
    pub name: String,
    pub kind: InterfaceKind,
    pub mac_address: Option<crate::types::MACAddress>,
    pub virtual_interface: FDURecordVirtualInterface,
    pub cp_uuid: Option<Uuid>, //internal to this descriptor
    pub intf_uuid: Uuid,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct FDURecordVirtualInterface {
    pub vif_kind: VirtualInterfaceKind,
    pub bandwidht: Option<u8>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct FDURecordConnectionPoint {
    pub uuid: Uuid,
    pub id: String,
}

// #[derive(Serialize,Deserialize,Debug, Clone)]
// pub struct FDUMigrationProperties {
//     pub destination : String,
//     pub source : String
// }

// #[derive(Serialize,Deserialize,Debug, Clone)]
// pub struct Record {
//     pub uuid : String,
//     pub fdu_id : String,
//     pub status : FDUState,
//     pub image : Option<Image>,
//     pub compute_requirements : ComputationalRequirements,
//     pub interfaces : Vec<FDURecordInterface>,
//     pub ssh_keys : Vec<String>,
//     pub hypervisor : String, //eg. Docker, KVM, LXD, ROS2, Native ...
//     pub migration_kind : MigrationKind,
//     pub geographic_requirement : Option<GeographicalRequirement>,
//     pub properties : Option<String>,
//     pub error_code : Option<u64>,
//     pub error_msg : Option<String>,
//     pub migration_properties : Option<FDUMigrationProperties>,
//     pub hypervisor_info : String,
// }
